streams/shared/src/main/scala/zio/stream/internal/Builder.scala

```diff
@@ -0,0 +1,114 @@
+package zio.stream.internal
+
+import zio._
+
+/**
+ * A mutable builder used for aggregating values of type `I` into values of type `O`
+ * through a [[ZSummary]].
+ */
+sealed abstract class Builder[-R, +E, -I, +O] { self =>
+
+  /**
+   * Wraps this builder as an effectual builder. If it is already
+   * an effectual builder, this operation does nothing.
+   */
+  def effectual: Builder.Effectual[R, E, I, O] =
+    self match {
+      case effectual: Builder.Effectual[R, E, I, O] => effectual
+      case mutable: Builder.Mutable[I, O] =>
+        new Builder.Effectual[Any, Nothing, I, O] {
+          override def reset: ZIO[Any, Nothing, Unit]            = UIO(mutable.reset())
+          override def accepts(i: I): ZIO[Any, Nothing, Boolean] = UIO(mutable.accepts(i))
+          override def done: ZIO[Any, Nothing, Boolean]          = UIO(mutable.done())
+          override def step(i: I): ZIO[Any, Nothing, Unit]       = UIO(mutable.step(i))
+          override def extract: ZIO[Any, Nothing, O]             = UIO(mutable.extract())
+        }
+    }
+}
+
+object Builder {
+
+  /**
+   * An impure, mutable builder that summarizes `I` values into `O` values. Values
+   * of this type should be considered mutable and unsafe to invoke concurrently.
+   */
+  abstract class Mutable[-I, +O] extends Builder[Any, Nothing, I, O] {
+
+    /**
+     * Resets the state of the builder. This should be equivalent to creating
+     * a fresh builder.
+     */
+    def reset(): Unit
+
+    /**
+     * Copies this builder to a new builder. The new builder must not reference any
+     * mutable state in this builder.
+     *
+     * The state of the resulting builder is undefined. It is required to call `reset`
+     * before stepping the builder.
+     */
+    def copy(): Mutable[I, O]
+
+    /**
+     * Tests whether the builder will accept the proposed element. The caller
+     * must always verify this predicate before calling `step`, and must
+     * only call `step` if this method returns `true`.
+     */
+    def accepts(i: I): Boolean
```

Comment:
Can you clarify the relationship between accepts and reset as you are using in for example
`ZTransducer#fromSummary?`

Reply:
I too feel like we've kind of gone backward a bit, in that we're incorporating a notion of "I'm done" into a summary that is never really done, in order to support aggregate async use case.